_posts/2018-12-06-pmreorder-basics.md

```diff
@@ -0,0 +1,318 @@
+---
+title: Pmreorder basics
+author: wlemkows
+layout: post
+identifier: pmreorder-basics
+---
+
+
+### Introduction
+
+
+When you write your persistent memory application the good practice is
+to check it under pmemcheck - the tool which we have already described in two
+blog posts some time ago. Feel free to come back to them and remind details
+about [basics usage] [pmemcheck1Link] and [transaction][pmemcheck2Link].
+
+In this post, I am going to introduce another tool for persistence correctness check.
+As you already know, pmemcheck verify if all stores are made persistent in a proper
+manner, pmreorder does not replace this functionality but provide extended one.
+Pmreorder traverse sequences of stores between flush-fence barriers made by the application.
+As the next step tool repeats those memory operations many times in different combinations
+to simulate possible order of stores on the NVDIMM.
+
+Each combination of stores is then verified by user-defined consistency checker and
+any inconsistencies are logged. This process allows detecting not persistent data
```

Comment:
Given an exhaustive consistency checking function, this process will uncover potential application bugs that otherwise could have been encountered only under a specific system failures.